lib/demuxeres/gstdemuxeres.c

```diff
@@ -0,0 +1,655 @@
+      break;
+    }
+    case DEMUXER_ES_STREAM_TYPE_AUDIO:
+    {
+      gst_audio_info_from_caps(&stream->data.audio.info, caps);
+      gst_structure_get_int (s, "bitrate", &stream->data.audio.bitrate);
+      stream->data.audio.acodec = gst_parse_stream_get_acodec_from_caps (caps);
+      break;
+    }
+    default:
+      break;
+  }
+  stream->id = g_list_length (priv->streams);
+
+  appsink = gst_element_factory_make ("appsink", NULL);
+  g_object_set (appsink, "sync", FALSE, "emit-signals", TRUE, NULL);
+
+  gst_bin_add_many (GST_BIN (priv->pipeline), appsink, NULL);
+  sinkpad = gst_element_get_static_pad (appsink, "sink");
+  if ((link = gst_pad_link (pad, sinkpad)) != GST_PAD_LINK_OK) {
+    GST_ERROR ("Unable to link the appsink with the parser link status %d",
+        link);
+    gst_parse_stream_teardown(stream);
+    stream = NULL;
+  }
+
+  GST_DEBUG_BIN_TO_DOT_FILE_WITH_TS (GST_BIN (priv->pipeline),
+      GST_DEBUG_GRAPH_SHOW_ALL, "gst-demuxerer.stream_create");
+
+  if (appsink)
+    gst_element_sync_state_with_parent (appsink);
+
+  callbacks.new_sample =appsink_new_sample_cb;
+  gst_app_sink_set_callbacks ((GstAppSink *) appsink, &callbacks, stream,
+      NULL);
+
+  gst_caps_unref (caps);
+  gst_object_unref (sinkpad);
+  return stream;
+}
+
+static inline gboolean
+wait_for_demuxer_ready (GstDemuxerES * demuxer, gint64 wait_time)
+{
+  GstDemuxerESPrivate *priv = demuxer->priv;
+  gboolean success = FALSE;
+  gint64 end_time = g_get_monotonic_time () + wait_time;
+
+  g_mutex_lock (&priv->ready_mutex);
+  while (!success) {
+
+    success = g_cond_wait_until (&priv->ready_cond,
+        &priv->ready_mutex, end_time);
+    GST_DEBUG ("After wait success %d", success);
+
+    priv->state = success ? DEMUXER_ES_STATE_READY : DEMUXER_ES_RESULT_ERROR;
+    success = TRUE;
+  }
+  g_mutex_unlock (&priv->ready_mutex);
+
+  return (priv->state == DEMUXER_ES_STATE_READY);
+}
+
+static void
+uridecodebin_pad_added_cb (GstElement * uridecodebin, GstPad * pad,
+    GstDemuxerES * demuxer)
+{
+  GstDemuxerEStream *stream = NULL;
+
+  if (!GST_PAD_IS_SRC (pad))
+    return;
+
+  GST_DEBUG ("pad %s:%s", GST_DEBUG_PAD_NAME (pad));
+
+  stream = gst_parse_stream_create (demuxer, pad);
+  if (!stream)
+    return;
+
+  demuxer->priv->streams = g_list_append (demuxer->priv->streams, stream);
+  GST_DEBUG ("Done linking");
+}
+
+static void
+uridecodebin_pad_no_more_pads (GstElement * uridecodebin,
+    GstDemuxerES * demuxer)
+{
+  GST_INFO ("No more pads received from %s", GST_ELEMENT_NAME (uridecodebin));
+}
+
+static GstAutoplugSelectResult
+uridecodebin_autoplug_select_cb (GstElement * src, GstPad * pad,
+    GstCaps * caps, GstElementFactory * factory, GstDemuxerES * demuxer)
+{
+  GstAutoplugSelectResult ret = GST_AUTOPLUG_SELECT_TRY;
+  if (gst_element_factory_list_is_type (factory,
+          GST_ELEMENT_FACTORY_TYPE_DEPAYLOADER
+          | GST_ELEMENT_FACTORY_TYPE_DECODER | GST_ELEMENT_FACTORY_TYPE_SINK)) {
```

Comment:
I haven't used this API, but for me, GST_ELEMENT_FACTORY_TYPE_DECODER should be the only type to check, the rest should be ignored.

Reply:
i'm wondering what's happening with a RAW media file.